### PR TITLE
REBASE: feat(mock-server)!: First full implementation

### LINKED ISSUE #264: Extend mock server to support integration testing
Extend the mock server to support integration testing.

### Useful functionalities:

- Function to send messages and reply to received messages
- Hooks to intercept receiving and sending messages
- Hooks to intercept specific types of messages, manipulate them 
- "Message store" which holds all received and sent messages and which is clearable via a function
- exposure of the sockjs and http server instance

Ideas and suggestions are welcome :wink:

### LINKED ISSUE #273: Document Vert.x mock server
Document the Vert.x mock server further.
See #264 for implementation specifications.

### PR DESCRIPTION
**Pull request by fussel178**
### Summary 

First complete implementation of the Vert.x mock server with improved abstractions and hooks.

### Details 

- new abstractions and hooks
- better/more documentation

#### BREAKING CHANGES:
- The listen function now requires an object as argument which contains the port and hostname as properties.
- The event bus and http server instance properties are now marked private. Please use the provided abstractions and hooks instead. (see onInit, send, handle, register, etc.)

### Related links 

- Fixes #264 
- Fixes #273 

### CLA

- [x] I have signed the individual contributor's license agreement **and sent** it to the board of the WüSpace e. V. organization.

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -26,5 +26,22 @@ module.exports = {
...
+		// 	},
+		// 	preset: 'ts-jest',
+		// 	testEnvironment: 'node',
+		// 	collectCoverageFrom: [
+		// 		'<rootDir>/packages/vertx-mock-server/src/**'
+		// 	],
+		// 	coveragePathIgnorePatterns: [],
+		// 	testMatch: [
+		// 		'<rootDir>/packages/vertx-mock-server/src/**/*(*.)@(spec|test).[jt]s'
+		// 	]
```
Comment (pklaschka) [inline jest.config.js:44]: Could you please add a `// TODO:` item here for what to do before uncommenting this section (adding an explanation of why it's commented out in the process)?

File: packages/vertx-mock-server/src/model/callback.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	message: JsonSerializable;
+	/**
+	 * The type of the received message.
+	 */
+	type: 'send' | 'publish';
+}
+
+/**
+ * The arguments of a send callback which gets called
+ * when the Vertx. mock server receives
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/model/callback.ts:20]:
```suggestion
 * when the Vert.x mock server receives
```

File: packages/vertx-mock-server/src/model/callback.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	/**
+	 * A callback to respond to the sent message.
+	 * @param response - the response message to send back
+	 */
+	respond: (response: JsonSerializable) => void;
+}
+
+/**
+ * The arguments of a publish callback which gets called
+ * when the Vertx. mock server receives
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/model/callback.ts:34]:
```suggestion
 * when the Vert.x mock server receives
```

File: packages/vertx-mock-server/src/model/extensions.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+ * when a new connection with a client is established
+ * and all {@link MockServer.register} calls has taken place.
+ *
+ * @see {@link MockServer}
+ */
+export interface OnConnection {
+	/**
+	 * An event handler which gets called
+	 * when a new connection with a client is established
+	 * and all {@link MockServer.register} calls has taken place.
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/model/extensions.ts:31]:
```suggestion
	 * and all {@link MockServer.register} calls have taken place.
```

File: packages/vertx-mock-server/src/model/extensions.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	 * when a the Vert.x mock server is set up
+	 * and ready to handle incoming connections.
+	 */
+	onInit(): void;
+}
+
+/**
+ * Defines an event handler that gets called
+ * when a new connection with a client is established
+ * and all {@link MockServer.register} calls has taken place.
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/model/extensions.ts:23]:
```suggestion
 * and all {@link MockServer.register} calls have taken place.
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 6]
```diff
@@ -1,39 +1,510 @@
...
+
+const logger = getLogger('Vert.x Mock Server');
+
+/**
+ * A mock server for a Vert.x WebSocket Event Bus bridge
+ * to JavaScript clients like {@link @wuespace/vertx-event-bus#EventBus}.
+ */
+export class MockServer {
+	/**
+	 * Holds all active connections with the Vert.x mock server.
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:34]:
```suggestion
	 * Holds all active connections of the Vert.x mock server.
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 7]
```diff
@@ -1,39 +1,510 @@
...
+		this.callbacks = [];
+
 		this.eventBus = sockjs.createServer();
 		this.httpServer = http.createServer();
 		this.setupEventBus(prefix);
 	}
 
-	listen(port = 9870, hostname = '0.0.0.0'): void {
+	/**
+	 * Starts the Vert.x mock server and listening for incoming connections.
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:95]:
```suggestion
	 * Starts the Vert.x mock server, listening for incoming connections.
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 8]
```diff
@@ -1,39 +1,510 @@
...
+	 * ```
+	 */
+	listen({ port = 9870, hostname = '0.0.0.0' }: ListenOptions): void {
 		this.httpServer.listen(port, hostname);
+		this.onInit?.();
+	}
+
+	/**
+	 * Stops the server from accepting new connections
+	 * and close currently active connections.
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:114]:
```suggestion
	 * Stops the server from accepting new connections
	 * and closes currently active connections.
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 9]
```diff
@@ -1,39 +1,510 @@
...
+	close(): Promise<void> {
+		this.onClose?.();
+		return new Promise<void>((resolve, reject) =>
+			this.httpServer.close(err => (err ? reject(err) : resolve()))
+		);
+	}
+
+	/**
+	 * Sends a message to the specified channel.
+	 * @param channel - the channel the message are sent to
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:137]:
```suggestion
	 * @param channel - the channel the message gets sent to
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 10]
```diff
@@ -1,39 +1,510 @@
...
+	 * @param message - the message that are sent over the specified channel
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:138]:
```suggestion
	 * @param message - the message that gets sent over the specified channel
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 11]
```diff
@@ -1,39 +1,510 @@
...
+	 * ```
+	 */
+	protected send(channel: ChannelAddress, message: JsonSerializable): void {
+		this.pushMessage(encodeReceiveMessage(channel, message));
+	}
+
+	/**
+	 * Sends an error message to the specified channel.
+	 * @param channel - the channel the message are sent to
+	 * @param error - the message that are sent over the specified channel
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:179]:
```suggestion
	 * @param channel - the channel the message gets sent to
	 * @param error - the message that gets sent over the specified channel
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 12]
```diff
@@ -1,39 +1,510 @@
...
+	 */
+	protected sendError(channel: ChannelAddress, error: ErrorContent): void {
+		this.pushMessage(encodeErrorMessage(channel, error));
+	}
+
+	/**
+	 * Registers a callback to a specified channel.
+	 * The callback gets called when a new message arrives on this channel.
+	 *
+	 * @param channel - the channel the callback is registered to
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:220]:
```suggestion
	 * @param channel - the channel the callback gets registered to
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 13]
```diff
@@ -1,39 +1,510 @@
...
+	 */
+	protected register(channel: ChannelAddress, callback: Callback): CallbackId {
+		return this.callbacks.push([channel, callback]);
+	}
+
+	/**
+	 * Unregisters/Removes a callback from a channel
+	 * with the given callback id from the {@link MockServer.register} function.
+	 *
+	 * @param id - the callback id from the {@link MockServer.register} function
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:257]:
```suggestion
	 * @param id - the callback id returned by the {@link MockServer.register} function
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 14]
```diff
@@ -1,39 +1,510 @@
...
+	 * const server = new MyMockServer();
+	 * server.listen();
+	 * ```
+	 */
+	protected getConnections(): Array<Connection> {
+		return [...this.connections];
 	}
 
+	/**
+	 * Pushes a raw/encoded message onto a active connections.
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:319]:
```suggestion
	 * Pushes a raw/encoded message onto all active connections.
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 15]
```diff
@@ -1,39 +1,510 @@
...
+	 */
+	private pushMessage(data: string): void {
+		this.connections.forEach(conn => conn.write(data));
+	}
+
+	/**
+	 * Set up the internal event bus, register all necessary event calls
+	 * for incoming connections and attaches to the internal http server instance.
+	 *
+	 * @param prefix - the prefix under which the event bus is reachable
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:340]:
```suggestion
	 * Set up the internal event bus, register all necessary event calls
	 * for incoming connections and attach it to the internal HTTP server instance.
	 *
	 * @param prefix - the prefix under which you can reach the event bus
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 16]
```diff
@@ -1,39 +1,510 @@
...
 	}
+
+	/**
+	 * Handles the incoming data, decodes them sorts them
+	 * and calls the registered callbacks.
+	 *
+	 * @param conn - the connection the message came from
+	 * @param data - the raw/encoded message
+	 *
+	 * @see {@link MockServer.setupEventBus}
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:377]:
```suggestion
	 * Handles the incoming data, decodes them, sorts them
	 * and calls the registered callbacks.
	 *
	 * @param conn - the connection the message came from
	 * @param data - the raw/encoded message
	 *
	 * @see {@link MockServer.setupEventBus}
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 17]
```diff
@@ -1,39 +1,510 @@
...
+	}
+
+	/**
+	 * Handles the closed event on the current connection.
+	 * It removes the given connection from the connection list
+	 * and call the connection closed hook.
+	 *
+	 * @param conn - the connection that has closed
+	 *
+	 * @see {@link MockServer.setupEventBus}
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:416]:
```suggestion
	 * Handles the closed event on the current connection.
	 * Removes the given connection from the connection list
	 * and calls the connection closed hook.
	 *
	 * @param conn - the closed connection
	 *
	 * @see {@link MockServer.setupEventBus}
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 18]
```diff
@@ -1,39 +1,510 @@
...
+	 */
+	private handleClose(conn: Connection): void {
+		this.connections = this.connections.filter(current => current !== conn);
+		this.onClosedConnection?.(conn);
+	}
+
+	/**
+	 * Processes the received message, determines the type,
+	 * builds the arguments for the callbacks
+	 * and call all registered callbacks on the message channel.
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:440]:
```suggestion
	 * Processes the received message, determines the type,
	 * builds the arguments for the callbacks
	 * and calls all registered callbacks on the message channel.
```

File: packages/vertx-mock-server/src/vertx-mock-server.ts [hunk 19]
```diff
@@ -1,39 +1,510 @@
...
+			if (channel === message.address) {
+				callback(args);
+			}
+		});
+	}
+
+	/**
+	 * Builds the respond handler for a 'send' message.
+	 * @param conn - the connection the message came from
+	 * @param message - the 'send' message to build the respond handler for
```
Comment (pklaschka) [inline packages/vertx-mock-server/src/vertx-mock-server.ts:476]:
```suggestion
	 * Builds the `respond` handler for a 'send' message.
	 * @param conn - the connection the message came from
	 * @param message - the 'send' message for which the `respond` handler gets built
```

**Review by pklaschka — CHANGES_REQUESTED** [review]
Overall, it looks good. I've left a few inline suggestions for some typos and other optimizations. 😉 

### One "major" request
Can we please make the logging optional? The solution doesn't have to be pretty, but I'd like to have the option of not polluting the console output in plugins and such.

**Comment by fussel178** [comment]
Now you can set a `ComponentLogger` in the constructor where the mock server and the SockJS server log it contents.
I also improved the current log output by a lot.

image

**Review by pklaschka — APPROVED** [review]
LGTM! 🎉 👍

**Pull request merged**